### PR TITLE
bpo-36287: Make ast.dump() not output optional fields and attributes.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Values for optional fields and attributes of AST nodes are now set as class attributes (e.g. Constant.kind is set to None).

https://bugs.python.org/issue36287

### CODE REVIEW

**Comment by pablogsal** [comment]
I think these attributes should still be printed if `include_attributes=True` in `ast.dump()`. For instance, we use this in the new PEG parser to quickly compare two full AST trees.

**Comment by pablogsal** [comment]
Suggestion: I think these attributes should still be printed if `include_attributes=True` in `ast.dump()`. For instance, we use this in the new PEG parser to quickly compare two full AST trees. It will still work (because we are only skipping them if they are not the default value) but I think it may make sense to have a "full verbose" mode.

What do you think?

File: Misc/NEWS.d/next/Library/2020-03-08-09-53-55.bpo-36287.mxr5m8.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+:func:`ast.dump()` no longer output optional fields and attributes. Values
```
Comment (pablogsal) [inline Misc/NEWS.d/next/Library/2020-03-08-09-53-55.bpo-36287.mxr5m8.rst]:
```suggestion
:func:`ast.dump()` no longer outputs optional fields and attributes. Values
```

File: Doc/library/ast.rst [hunk 2]
```diff
@@ -139,9 +139,10 @@ Literals
 
    .. doctest::
 
+        >>> import ast
```
Comment (pablogsal) [inline Doc/library/ast.rst]: This should be necessary for the doctest as we have

```
   .. testsetup::
    
        import ast
```

at the beggining of the document. Maybe I am missing something, though

File: Misc/NEWS.d/next/Library/2020-03-08-09-53-55.bpo-36287.mxr5m8.rst (hunk 1 above)
Comment (pablogsal) [reply Misc/NEWS.d/next/Library/2020-03-08-09-53-55.bpo-36287.mxr5m8.rst]: I think this explanation is slightly confusing. If I understand correctly is the **default** values what are now class attributes. For instance, for a non-default "kind":

```
>>> x = ast.parse("u'dsf'", mode="eval")
>>> x.body.kind
'u'
```

while the default for `kind` is what is set on the class:

```
>>> print(type(x.body).kind)
None
```

Maybe something like:

```
:func:`ast.dump()` no longer outputs optional fields and attributes with default values. The
default values for optional fields and attributes of AST nodes are now set as class
attributes (e.g. ``Constant.kind`` is set to ``None``).
```

File: Doc/library/ast.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/library/ast.rst]: I am not experienced with doctests. I ran `./python -m doctest Doc/library/ast.rst` and it failed. Seems `doctest` does not understand `testsetup`. Indeed, it is a Sphinx extension.

**Comment by serhiy-storchaka** [comment]
Currently the only attributes are "lineno", "col_offset", "end_lineno" and "end_col_offset". They have type `int`, the latter two are optional. When convert AST nodes from C to Python these attributes are always set to integer value (at worst case it is `0`). They are never None. When convert from Python to C None will be converted to `0`.

Currently if you create an AST node manually and not set attributes, they will not be output by `ast.dump()`.
```
>>> print(ast.dump(ast.Constant(), include_attributes=True))
Constant()
```
If always output attributes if `include_attributes=True`, you would get
```
>>> print(ast.dump(ast.Constant(), include_attributes=True))
Constant(end_lineno=None, end_col_offset=None)
```
Is it what you want to get?

**Comment by pablogsal** [comment]
> Is it what you want to get?

Sort of. What I am proposing (still don't know if its a good idea) is to have a way of "print everything, no matter if is default or not"). Feel free to ignore this if you don't see much value, I just wanted to know if you think it may make sense to preserve that behaviour under some option

File: Doc/library/ast.rst (hunk 2 above)
Comment (pablogsal) [reply Doc/library/ast.rst]: Yeah, you need to do "make doctest" in `Docs` or `PATH=./venv/bin:$PATH sphinx-build -b doctest -d build/doctrees   -W . build/doctest FILENAME`  if you want to only test one single file (the first run will be a bit longer while it builds the tree).

**Review by pablogsal — APPROVED** [review]
LGTM

I think this is a great improvement! Feel free to ignore the remaining suggestions if you don't think is worth tackling those.

**Comment by serhiy-storchaka** [comment]
I experimented and I do not like the result. It would break an existing test `test_dump_incomplete` which is not touched by this PR. It is not important. I think that in practical cases there will be not difference because you will always set instance attributes, and they are always different from default values (because they are ints, and the default is None).

In any case, this is just a first step, and there is a place for improvements. This change fixes the most annoying thing, `kind=None` in `Constant`, and the rest is less important.

File: Lib/ast.py [hunk 3]
```diff
@@ -123,31 +123,36 @@ def _format(node, level=0):
...
-                        value = getattr(node, attr)
+                        value = getattr(node, name)
                     except AttributeError:
-                        pass
-                    else:
-                        value, simple = _format(value, level)
-                        allsimple = allsimple and simple
-                        args.append('%s=%s' % (attr, value))
+                        continue
+                    if value is None and getattr(cls, name, ...) is None:
```
Comment (0dminnimda) [inline Lib/ast.py:151]: Do we need to check this twice (additional `getattr(cls, name, ...) is None`)? It seems like those `is None`s will always be the same

Comment (serhiy-storchaka) [reply Lib/ast.py:151]: Change and test.

Comment (0dminnimda) [reply Lib/ast.py:151]: Sorry, although I did not change, now I can see that I overlooked that this is `cls` and not `node`

Comment (0dminnimda) [reply Lib/ast.py:151]: this conversation can be resolved, I think

**Pull request merged**